CorsixTH/Lua/earthquake.lua

```diff
@@ -0,0 +1,240 @@
+    -- Mark the planned earthquake as ready to start in the next tick.
+    self.active = true
+  end
+end
+
+--! Called when it is time to have another earthquake
+function Earthquake:nextEarthquake()
+  self.active = false
+
+  local level_config = self.world.map.level_config
+  -- Check carefully that no value that we are going to use is going to be nil
+  if level_config.quake_control and level_config.quake_control[self.current_map_earthquake] and
+      level_config.quake_control[self.current_map_earthquake].Severity ~= 0 then
+    -- This map has rules to follow when making earthquakes, let's follow them
+    local control = level_config.quake_control[self.current_map_earthquake]
+    self.start_month = math.random(control.StartMonth, control.EndMonth)
+
+    -- Month length of the start of the earthquake. From start to finish
+    -- earthquakes do not persist for >= a month so we can wrap all days
+    -- after the start around the month length unambiguously.
+    local eqml = month_length[(self.start_month % 12) + 1]
+    self.start_day = math.random(1, eqml)
+
+    self.size = control.Severity
+    self.remaining_damage = self.size
+    self.damage_timer = damage_time
+    self.warning_timer = warning_period
+    self.current_map_earthquake = self.current_map_earthquake + 1
+    self.next_planned = true
+  end
+end
+
+--! Create a small or medium quake now for the "Create Earthquake" cheat
+function Earthquake:createEarthquake()
+  -- Make sure an earthquake isn't already happening, or are disabled
+  if self.active or self.disabled then return false end
+
+  self.start_day = self.world.game_date:dayOfMonth()
+  self.start_month = self.world.game_date:monthOfGame()
+  if not self.next_planned then
+    -- Forcefully make an earthquake if there are none left in level file
+    self.size = math.random(1, 6) -- Above 6 seems disastrous
+    self.remaining_damage = self.size
+    self.damage_timer = damage_time
+    self.warning_timer = warning_period
+  end
+end
+
+--! Check if there is a current earthquake
+--!return (boolean) True if there is a current earthquake
+function Earthquake:isActive()
+  return self.active
+end
+
+--! Afterload sections before 189 were taken from World
+--!param old (integer) The old version of the save game.
+--!param new (integer) The current version of the save game format.
+function Earthquake:afterLoad(old, new) -- luacheck: ignore 212 keep params from parent function
+  local old_quake = self.world.next_earthquake
+  if old < 115 then
+    if old_quake.active then
+      local rd = 0
+      for _, room in pairs(self.world.rooms) do
+        for object, _ in pairs(room.objects) do
+          if object.quake_points then
+            rd = math.max(rd, object.quake_points)
+            object.quake_points = nil
+          end
+        end
+      end
+      old_quake.remaining_damage = rd
+      old_quake.damage_timer = damage_time
+      old_quake.warning_timer = 0
+    end
+  end
```

Comment:
This is from 0.61... I don't think we would be at any risk of removing this afterload too.

Reply:
Now it's done, I'll leave it for a mass clean up.